### PR TITLE
Tournaments refactor

### LINKED ISSUE #365: Refactor the tournament logic

### PR DESCRIPTION
**Pull request by matt989253**
Closes #365

### CODE REVIEW

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/StoreDocumentTournament.kt [hunk 1]
```diff
@@ -134,11 +114,83 @@ class StoreDocumentTournament(
...
+                roundDepth = depth,
+                tournamentId = document.uid,
+            )
+        set(matchRef, matchDocument)
+      }
     }
   }
 
-  private suspend fun createMatchesForRankedPlayers(
+  private fun Transaction<DocumentReference>.createMatchesForPoolResults(
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/StoreDocumentTournament.kt:193]: Missing some documentation for this method.


```kotlin
/**
 * Creates the matches corresponding to the given [PoolResults].
 *
 * @receiver the ongoing [Transaction].
 * @param results the [PoolResults] for which the matches are created.
 * @param nextDepth the function which computes the depth of the elimination tree.
 */
private fun Transaction<DocumentReference>.createMatchesForPoolResults()
```

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/StoreDocumentTournament.kt [hunk 2]
```diff
@@ -134,11 +114,83 @@ class StoreDocumentTournament(
...
+              tournamentBestOf = bestOf,
+              tournamentAdminId = adminId,
+              playerIds = participants.map { it.first },
+              playerNames = participants.map { it.second },
+          )
+        }
+        .forEach { set(store.collection(PoolDocument.Collection).document(), it) }
+  }
+
+  private fun Transaction<DocumentReference>.createFinalsMatchesForPlayers(
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/StoreDocumentTournament.kt:158]: Missing documentation :


```kotlin
/**
 * Creates the matches corresponding to the given finals phase.
 *
 * @receiver the ongoing [Transaction].
 * @param players the identifiers of the players, 
 * @param nextDepth the function which computes the depth of the elimination tree.
 */
private fun Transaction<DocumentReference>.createFinalsMatchesForPlayers()
```

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/StoreDocumentTournament.kt [hunk 3]
```diff
@@ -134,11 +114,83 @@ class StoreDocumentTournament(
...
+
+  private fun Transaction<DocumentReference>.createPools(
+      tournamentDocumentReference: DocumentReference,
+      current: TournamentDocument,
+      maxPlayers: Int,
+      poolSize: Int,
+      bestOf: Int,
+      adminId: String,
+      players: List<String>
+  ) {
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/StoreDocumentTournament.kt:129]: Missing documentation.

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/StoreDocumentTournament.kt (hunk 1 above)
Comment (matt989253) [reply mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/StoreDocumentTournament.kt:220]: I did not know that we needed to document private functions?
Anyway, fixed in https://github.com/epfl-SDP/android/pull/392/commits/a8685cd7be6104b3a4f8315aabff2c43f026a76b

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/StoreDocumentTournament.kt (hunk 2 above)
Comment (matt989253) [reply mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/StoreDocumentTournament.kt:179]: Fixed in https://github.com/epfl-SDP/android/pull/392/commits/a8685cd7be6104b3a4f8315aabff2c43f026a76b

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/StoreDocumentTournament.kt (hunk 3 above)
Comment (matt989253) [reply mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/StoreDocumentTournament.kt:143]: Anyway, fixed in https://github.com/epfl-SDP/android/pull/392/commits/a8685cd7be6104b3a4f8315aabff2c43f026a76b

**Review by alexandrepiveteau — APPROVED** [review]

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/StoreDocumentTournament.kt [hunk 4]
```diff
@@ -60,56 +57,39 @@ class StoreDocumentTournament(
...
+        val adminId = current.adminId ?: return@transaction
+        val maxPlayers = current.maxPlayers ?: return@transaction
+        val poolSize = current.poolSize ?: return@transaction
+        val players = current.playerIds ?: return@transaction
+        val bestOf = current.bestOf ?: return@transaction
+        val stage = current.stage
+
+        // Are we the tournament admin ?
+        if (adminId != user.uid) return@transaction
+        if (stage != null) return@transaction
```
Comment (KurohanaJuri) [inline mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/StoreDocumentTournament.kt:74]: Can you replace this by `val stage = current.stage ?: return@transaction` at line 70? Or they haven't the same meaning ?

File: mobile/src/androidTest/java/ch/epfl/sdp/mobile/test/state/Dsl.kt [hunk 5]
```diff
@@ -146,6 +150,16 @@ suspend fun ComposeContentTestRule.setContentWithTestEnvironment(
   return environment
 }
 
+/**
+ * Wait until a certain text is visible before clicking on it
+ *
+ * @param text the text
```
Comment (KurohanaJuri) [inline mobile/src/androidTest/java/ch/epfl/sdp/mobile/test/state/Dsl.kt:156]:
```suggestion
 * @param text the expected text
```

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/StoreDocumentTournament.kt (hunk 4 above)
Comment (matt989253) [reply mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/StoreDocumentTournament.kt:74]: No, we actually need the opposite of what the elvis notation (`?:`) offers, we need `stage` to be `null`.

**Review by alexandrepiveteau — APPROVED** [review]

**Review by KurohanaJuri — APPROVED** [review]

**Pull request merged**